### PR TITLE
feat ✨: configurable idle timeout for GPU model release

### LINKED ISSUE #70: Feature request: configurable idle timeout for GPU model release
## Summary

Add a user-configurable idle timeout for releasing cached GPU ASR models.

## Current behavior

`IFW_EAGER_MODEL_RELEASE=1` releases the backend immediately when a transcription finishes. With the variable unset, the model remains warm in the process cache indefinitely.

## Requested behavior

Introduce a timeout environment variable and corresponding configuration constant, for example `IFW_MODEL_RELEASE_TIMEOUT_SECONDS`.

- Unset or disabled: retain the current warm-cache behavior.
- `0`: preserve immediate release semantics (equivalent to `IFW_EAGER_MODEL_RELEASE=1`).
- Positive value: release the cached backend only after it has had no active borrowers for the configured idle period.
- A new request before the timeout expires should keep/reuse the loaded model and reset/cancel the pending release.

This would let users balance GPU VRAM availability against first-request latency without choosing only between indefinite caching and immediate model reloads.

## Context

A persistent WebUI process with `openai/whisper-large-v3-turbo` retains roughly 2 GiB of ROCm VRAM after work is complete.

### PR DESCRIPTION
**Pull request by beecave-orchestrator**
# Pull Request: Configurable idle timeout for GPU model release

## Summary

Add `IFW_MODEL_RELEASE_TIMEOUT_SECONDS` to let users release cached GPU ASR models after a configurable idle period, balancing VRAM availability against first-request latency.

## Why

- Issue #70: the existing `IFW_EAGER_MODEL_RELEASE=1` only offers immediate release or indefinite warm caching. A persistent WebUI with `whisper-large-v3-turbo` retains ~2 GiB ROCm VRAM after work completes.
- Users need a middle ground: release the model after N seconds of idleness so short bursts of requests reuse the warm model while long gaps free VRAM.

## Notable changes (reviewer focus)

- **`constants.py`**: Parse `IFW_MODEL_RELEASE_TIMEOUT_SECONDS` with strict validation. Unset/blank/malformed/negative/non-finite values resolve to `None` (disabled, warm cache). `0` means immediate release. Positive values schedule a delayed release.
- **`backend_cache.py`**: `_CacheEntry` gains `_release_timer` and `_release_generation` fields. `release_pipeline` now follows a 3-way policy: eager (immediate, overrides timeout), `None` (warm), `0` (immediate), positive (schedule timer). The timer callback validates generation + refcount under `_LOCK`, detaches the entry, then closes the backend *outside* the lock to avoid blocking.
- **`acquire_pipeline`** cancels pending timers and bumps the generation, so a reacquire race invalidates any previously scheduled release.
- **`clear_cache` and `invalidate_gpu_cache`** cancel pending timers before closing backends (forceful paths preserved).
- **`.env.example`**: Documented the new variable with semantics and defaults.

## Files changed

- **Counts:** 4 files changed (3 modified, 1 added).
- **Key touchpoints:** `constants.py` (config parsing), `backend_cache.py` (timer lifecycle), `.env.example` (docs), `test_backend_cache_timeout.py` (new tests).

<details>
<summary>File lists (auto)</summary>

### Added
- `tests/core/test_backend_cache_timeout.py`

### Modified
- `insanely_fast_whisper_rocm/utils/constants.py`
- `insanely_fast_whisper_rocm/core/backend_cache.py`
- `.env.example`

### Deleted
- (none)

</details>

## Test plan

- [x] Unit: 25 new deterministic tests with fake `threading.Timer` covering constants parsing (12 parametrized cases), immediate/zero release, delayed timer scheduling, timer fires + close, reacquire cancels timer, stale-timer generation guard, eager precedence, warm-cache (None), borrow_pipeline timer, clear_cache cancels timer, invalidate_gpu_cache cancels timer, missing-entry and nonzero-refcount edge cases.
- [x] Regression: all 11 existing `test_backend_cache.py` tests pass unchanged.
- [x] Lint: `ruff check` and `ruff format --check` pass on all changed files.
- [ ] Integration: deferred (requires GPU/ROCm runtime; no GPU deps installed in CI container).

## Risks & rollback

- **Risks:** Timer-based release runs in a daemon thread; the generation guard prevents stale-timer races, but under extreme contention a close could race with a new acquire. Mitigated by validating under `_LOCK` and closing outside it. Default behavior (timeout unset) is unchanged.
- **Rollback:** Unset `IFW_MODEL_RELEASE_TIMEOUT_SECONDS` or revert this PR. No data migration required.

## Additional notes

- Closes #70.
- 3 pre-existing test failures (`test_srt_quality` x2, `test_env_loader` x1) are unrelated to this change and reproduce on clean `main`.
- GPU/ROCm/PyTorch runtime not available in the CI container; focused unit tests use mocks and a fake timer to verify all timer lifecycle paths deterministically.

## Summary by CodeRabbit

* **New Features**
  * Added configurable model release timing through the `IFW_MODEL_RELEASE_TIMEOUT_SECONDS` setting.
  * Supports immediate, delayed, eager, or indefinite model retention.
  * Reacquiring a model cancels pending releases, helping prevent unnecessary reloads.

* **Bug Fixes**
  * Improved cache clearing and GPU invalidation to safely cancel scheduled releases and avoid stale model cleanup.

### CODE REVIEW

File: insanely_fast_whisper_rocm/core/backend_cache.py [hunk 1]
```diff
@@ -77,6 +86,65 @@ def _make_key(
...
+        entry: The cache entry to release after the timeout.
+        timeout: Delay in seconds before the backend is closed.
+    """
+    _cancel_release_timer(entry)
+    entry._release_generation += 1
+    generation = entry._release_generation
+    timer = threading.Timer(timeout, _timed_release, args=(key, generation))
+    timer.daemon = True
+    entry._release_timer = timer
+    timer.start()
```
[the remark this answers is not in the document]
Comment (beecave-orchestrator) [reply insanely_fast_whisper_rocm/core/backend_cache.py:128]: Fixed in b324abe8772763cb2afafafdd3f21b0634e8046b. Release generations are now process-wide monotonic values, used for both scheduling and reacquisition. Added a deterministic recreated-entry regression test: a cancelled/in-flight stale timer from the old entry cannot close a newly recreated idle backend with the same cache key.

File: tests/core/test_backend_cache_timeout.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+# ---------------------------------------------------------------------------
+
+
+class TestTimeoutRelease:
+    """Test idle-timeout release behavior in the backend cache."""
+
+    def test_timeout_zero_releases_immediately(self) -> None:
+        """Timeout=0 closes the backend immediately on release."""
+        backend_cache._EAGER_RELEASE = False
+        backend_cache._RELEASE_TIMEOUT = 0.0
```
[the remark this answers is not in the document]
Comment (beecave-orchestrator) [reply tests/core/test_backend_cache_timeout.py]: Fixed in b324abe8772763cb2afafafdd3f21b0634e8046b. Replaced every timeout-test global assignment with the release_policy fixture, which uses monkeypatch.setattr and is automatically restored. Added focused coverage confirming the overrides are registered for restoration.
